asf_heat_pump_suitability/pipeline/suitability/calculate_suitability.py

```diff
@@ -0,0 +1,472 @@
+
+    Returns:
+        pl.DataFrame: weighted scores for an LSOA
+    """
+    score_cols = [col for col in df.columns if "score" in col]
+    df = df.with_columns(
+        pl.when(
+            (pl.col("proportional_weight").is_not_null().sum() / len(df)) >= threshold
+        )
+        .then(pl.col("proportional_weight") / pl.col("proportional_weight").sum())
+        .otherwise(1)
+        .alias("use_weight"),
+        pl.when(
+            (pl.col("proportional_weight").is_not_null().sum() / len(df)) >= threshold
+        )
+        .then(True)
+        .otherwise(False)
+        .alias("scores_weighted"),
+    )
+    for col in score_cols:
+        df = df.with_columns(
+            (pl.col(col) * pl.col("use_weight")).alias(f"{col}_weighted")
+        )
+
+    return df
+
+
+def compute_dict_lsoa_suitability_scores(df: pl.DataFrame, lsoa: str) -> dict:
+    """
+    Calculate average heat pump suitability scores for LSOA per tech type.
+
+    Args:
+        df (pl.DataFrame): LSOA with weighted suitability scores per tech type
+        lsoa (str): LSOA code
+
+    Returns:
+        dict: suitability scores per LSOA for each tech type
+    """
+    scores_dict = {"lsoa": lsoa}
+    assert df["scores_weighted"].n_unique() == 1
+    score_cols = [col for col in df.columns if "score_weighted" in col]
+    for score in score_cols:
+        if df["scores_weighted"].unique()[0]:
+            scores_dict[score] = df[score].sum()
+        else:
+            scores_dict[score] = df[score].mean()
+    scores_dict["scores_weighted"] = df["scores_weighted"].unique()[0]
+
+    return scores_dict
+
+
+if __name__ == "__main__":
+    # TODO: logging.info not displaying to terminal for me
+    logging.info("Loading EPC data with features")
+    epc_df = get_enhanced_epc()
+
+    logging.info("Filtering EPC data to rows with n_features >= minimum threshold")
+    epc_df = filter_df_minimum_features(epc_df)
+
+    tech_types = [
+        "ASHP_S",
+        "ASHP_N",
+        "GSHP_S",
+        "GSHP_N",
+        "SGL_S",
+        "SGL_N",
+        "HN_S",
+        "HN_N",
+    ]
+
+    scores = []
+    for tech_type in tech_types:
+        logging.info(f"Calculating suitability scores for tech type: {tech_type}")
+        epc_scores_df = compute_df_avg_score_per_epc(epc_df, tech_type)
+        scores.append(epc_scores_df)
+
+    logging.info("Joining all scores to EPC dataset")
+    for score_df in scores:
+        epc_df = epc_df.join(score_df, on="UPRN", how="left")
+
+    fs = s3fs.S3FileSystem()
+    save_as = f"s3://asf-heat-pump-suitability/outputs/{datetime.today().strftime('%Y%m%d')}_2023_Q4_heat_pump_suitability_per_property.parquet"
+    with fs.open(save_as, mode="wb") as f:
+        epc_df.write_parquet(f)
+
+    logging.info("Weighting scores and aggregating per LSOA")
+    weighted_scores = []
+    for lsoa_code in tqdm(epc_df["lsoa"].unique()):
+        lsoa_df = epc_df.filter(pl.col("lsoa") == lsoa_code)
```

Comment:
This is a time for groupbys (no need to change though)
```suggestion
   for lsoa_code, lsoa_df in tqdm(epc_df.group_by("lsoa")): 
```

Reply:
I was using this originally (because that was also in your original code) but for some reason `lsoa_code` was returning as a tuple in this format: `('S01008901',)`. Couldn't figure out why and then just thought it's clearer to change to filter than do `lsoa_code[0]`.

Can come back to it in refactoring. Do you think group_by will be significantly quicker?

Reply:
ah I see.. maybe a polars format thing that I don't know about. It might be a bit quicker, but don't have a good sense of how much. Perhaps not worth changing then